### ISSUE TITLE
When fetching git+ssh flake inputs, prompt to enter SSH key passphrase is invisble, making it appear to hang

### ISSUE THREAD

**Issue by cumber** [NixOS/nix#8305]
**Describe the bug**

When Nix is fetching a `git+ssh` input for a flake, it will use my SSH agent and `~/.ssh/config` to authenticate. If the necessary SSH key is already in the agent, or doesn't have a passphrase, things work fine.

However it seems that when a passphrase is needed SSH reads the passphrase from stdin, but the prompt that SSH displays is hidden by Nix! This makes it seem like the process has simply hung, when in fact it is waiting for the user to enter a passphrase. (It also means you can't see which key it's using in order to know which passphrase to enter, if there was more than one option)

**Steps To Reproduce**

1. Have an SSH key for github.com that is protected by a passphrase
2. `git clone git@github.com:cumber/flake-with-ssh-input` (a minimal example flake)
3. `cd flake-with-ssh-input`
4. `ssh-agent -D` to clear the agent, if you use one
3. `nix flake lock --update-input trivial` (or any other command that will make Nix fetch the input)
5. The terminal will display `fetching Git repository 'ssh://git@github.com/cumber/flake-trivial.git` and then wait for (unechoed) input
6. Type your github.com SSH key passphrase, and the operation will complete
7. Alternatively, if you just hit enter (or otherwise enter an incorrect passphrase), an error like this will be displayed, with little hint that the reason is that it needed a passphrase to load an SSH key
    ```
   git@github.com: Permission denied (publickey).
   fatal: Could not read from remote repository.

   Please make sure you have the correct access rights
   and the repository exists.
   ```

The same thing also happens with things like `nix build git+ssh://git@github.com/cumber/trivial-flake`, so it's not strictly a problem with flake inputs as such, just fetching a `git+ssh` flake url. However when referenced this way Nix seems to cache the repo once you've fetched it successfully once and then the build succeeds with the cached version on subsequent attempts, making it harder to repeatedly reproduce.

**Expected behavior**

When fetching a `git+ssh` url, the `Enter passphrase for key '...':` prompt should be displayed as it usually is when Git is fetching over SSH and a passphrase is required

**`nix-env --version` output**
> nix-env (Nix) 2.13.3

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by knedlsepp** [comment]
Seems like this is a duplicate of https://github.com/NixOS/nix/issues/7959

**Comment by cumber** [comment]
> Seems like this is a duplicate of #7959, which was recently fixed by #8018.

Aha! That's similar, but seems to be different. That pull request is included in 2.15.0 so I tested that. I can confirm that it allows the passphrase prompt to appear when doing `nix copy` (to or from an `ssh://` url), but flake refs that happen to be `git+ssh` type don't seem to go through the same code path.

I would assume the ability added in that pull request to pause/resume the progress bar writes would be useful in fixing this issue too.

I wonder; is the difference that here Nix is calling **Git**, and it's Git that is calling SSH? Wrapping pause/resume around places where SSH is called directly by Nix wouldn't help when Nix calls Git which calls SSH. That would be awkward if it's desired to clobber Git's output, but not SSH's.

**Comment by thufschmitt** [comment]
Indeed, #8018 only fixes the direct calls to `ssh` (by calling it in such a way that we can detect when the connection is established and disable the progress bar until then). I'm not sure how easily that could be adapted to the `git+ssh` case :thinking: @balsoft do you have an idea?
